agent/helm: Deprecate --kpr=partial|strict|disabled and use --kpr=true|false instead

brb opened the pull request:
See per commit.

Once the PR is merged, the following will be updated before the v1.14 release:

- [x] KPR GSG docs
- [ ] Upgrade guide
- [x] GH workflows update

brb committed:
agent: Add --kube-proxy-replacement=true|false

This commit extendss the --kube-proxy-replacement flag by adding two
additional parameters - "true" and "false". The previous parameters
"strict", "partial", and "disabled" are deprecated, and scheduled for
the removal in v1.15 or v1.16.

The "true" parameter is the same as "strict", while "false" is the same
as "partial". The main idea is that the latter allows a user to enable
each subfeature. For example, --kpr="false" with
--enable-node-port="true" enables the BPF NodePort. This should reduce a
mental strain which comes when trying to understand the flag.

Disabling each subfeature flag will be done in a subsequent commit.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
agent: Change flag defaults of ExternalIPs and HostPort

Disable both by default, as the previous commit made
--kube-proxy-replacement to default to "false", which means that after
the upgrade to v1.14.0 users would enabled both previously disabled
features.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
helm: Add --kube-proxy-replacement=boolean

One notable change is that we feed kubeProxyReplacement through
toString. This is because users might use both --set and --set-string to
set the param. The former with "true" or "false" as the value implicitly
converts the variable in Helm to boolean which makes the string
comparisons in Helm to fail.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
api/daemon: Extend KPR status to show new states

Signed-off-by: Martynas Pumputis <m@lambda.lt>

sayboras commented on daemon/cmd/kube_proxy_replacement_test.go:
```diff
@@ -185,11 +185,11 @@ func (s *KPRSuite) TestInitKubeProxyReplacementOptions(c *C) {
-		// KPR strict + no conntrack ipt rules: error, needs bpf masquerade
+		// KPR true + no conntrack ipt rules: error, needs bpf masquerade
 		{
 			"kpr-strict+no-conntrack-ipt-rules+no-bpf-masquerade",
```
nit: here as well?

brb replied:
Good catch! Updated.

sayboras commented on daemon/cmd/daemon_main.go:
```diff
@@ -327,7 +327,7 @@ func initializeFlags() {
 	option.BindEnv(Vp, option.EnablePolicy)
 
-	flags.Bool(option.EnableExternalIPs, defaults.EnableExternalIPs, fmt.Sprintf("Enable k8s service externalIPs feature (requires enabling %s)", option.EnableNodePort))
+	flags.Bool(option.EnableExternalIPs, false, fmt.Sprintf("Enable k8s service externalIPs feature (requires enabling %s)", option.EnableNodePort))
```
do we need to mention upgrade note for these two flags ?

brb replied:
Aye, I am planning to do in the follow up PR to update the docs (I want to get this PR merged before the freeze :grin: )

sayboras reviewed (approved):
LGTM, some nits as per below

brb committed:
gh/workflows: Use --kpr=true instead of --kpr=strict

Also, use --kpr=true in tests-smoke.yaml, as the socket LB issues on GHA
runners were resolved long time ago.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

nathanjsweet commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -227,7 +227,7 @@ cilium-agent [flags]
       --k8s-watcher-endpoint-selector string                    K8s endpoint watcher will watch for these k8s endpoints (default "metadata.name!=kube-scheduler,metadata.name!=kube-controller-manager,metadata.name!=etcd-operator,metadata.name!=gcp-controller-manager")
       --keep-config                                             When restoring state, keeps containers' configuration in place
-      --kube-proxy-replacement string                           Enable only selected features (will panic if any selected feature cannot be enabled) ("partial"), or enable all features (will panic if any feature cannot be enabled) ("strict"), or completely disable it (ignores any selected feature) ("disabled") (default "partial")
+      --kube-proxy-replacement string                           Enable only selected features (will panic if any selected feature cannot be enabled) ("false"), or enable all features (will panic if any feature cannot be enabled) ("true") (default "false")
```
"true" and "false" don't really seem to be the right words to use. Why not just keep "partial" and "strict"?

jrajahalme replied:
The problem is that the new `false` is not the same as the old `partial`. The change is that with false, host-port, external-ips, etc. need to be explicitly enabled, while with `partial` they were enabled by default.

Language-wise, `strict` and `partial` take much more explaining than `true` and `false`. Moreover, `false` here should be read to mean "full kube-proxy replecement has not been enabled"?

nathanjsweet reviewed (changes requested):
Everything looks good, but the "true", "false" options don't make semantic sense. "False" corresponds with the idea of "disabled" rather than "partial". Can we just keep the original words while eliminating the "disabled" option?
